### PR TITLE
Feature/permission dsl

### PR DESCRIPTION
**Pull request by antonivanopoulos**
Adds a permissive DSL method, `allow_paginative_on`, to the model extension to whitelist pagination fields

By using `allow_paginative_on`, we can specify fields that we want to allow pagination on for the model via requests from the client, rather than opening up everything. This change is also two-fold, in that the method can take in hash items, already mapped to a table and column, allowing for pagination on a `JOIN`'d table. By default, the specified name will be mapped to the classes table, with the specified name.

i.e.
```
allow_paginative_on :name	 					#=> { name: 'my_model_table.name }
allow_paginative_on name: 'other_model_table.name' 	#=> { name: 'other_model_table.name }
```

### CODE REVIEW

File: lib/paginative/models/model_extension.rb [hunk 1]
```diff
@@ -23,24 +26,88 @@ def self.with_id_from(id=0, limit=25)
...
-            return self.order(sanitized_ordering(self.table_name, field, order)).where("#{self.table_name}.#{primary_sort_field} <= ? AND (#{self.table_name}.#{primary_sort_field} != ? OR #{self.table_name}.#{secondary_sort_field} < ?)", primary_value, primary_value, secondary_value) if order.try(:downcase) == "desc"
-
-            self.order(sanitized_ordering(self.table_name, field, order)).where("#{self.table_name}.#{primary_sort_field} >= ? AND (#{self.table_name}.#{primary_sort_field} != ? OR #{self.table_name}.#{secondary_sort_field} > ?)", primary_value, primary_value, secondary_value)
+
+        fields = Array.wrap(field).flatten
+        values = Array.wrap(value).flatten
+        zipped = fields.zip(values)
+        fields, values = prune_fields(zipped).transpose
+
+        q = self.all
```
Comment (antonivanopoulos) [inline lib/paginative/models/model_extension.rb]: Probably should not be `all`, and should be `scoped` instead.

File: lib/paginative/models/model_extension.rb [hunk 2]
```diff
@@ -23,24 +26,88 @@ def self.with_id_from(id=0, limit=25)
...
+        if fields.present? && fields.any?
+          return raise "Too many" unless fields.length <= 2
+          return raise "Something...." unless values.length == fields.length
+
```
Comment (antonivanopoulos) [inline lib/paginative/models/model_extension.rb]: Need to change these back to something useful...think they need to have coverage as well, actually.

File: Gemfile [hunk 3]
```diff
@@ -1,21 +1,8 @@
-source "https://rubygems.org"
+source 'https://rubygems.org'
 
-# Declare your gem's dependencies in paginative.gemspec.
+# Declare your gem's dependencies in dre.gemspec.
```
Comment (RustComet) [inline Gemfile]: I'm guessing this was a result of copying the gem file over from dre for testing different rails versions

File: lib/paginative/models/ordering_helpers.rb [hunk 4]
```diff
@@ -3,20 +3,13 @@ module OrderingHelpers
     extend ActiveSupport::Concern
 
     included do
-      def self.sanitized_ordering(table_name, field, order)
-        if field.is_a? Array
-          return raise "Wrong number of sorting fields. Expected 2, got #{field.length}. If you want to sort by a singular field please pass field argument as a string rather than an array." unless field.length == 2
-
-          "#{table_name}.#{sanitize_column(field[0])} #{sanitize_column_direction(order)}, #{table_name}.#{sanitize_column(field[1])} #{sanitize_column_direction(order)}"
```
Comment (RustComet) [inline lib/paginative/models/ordering_helpers.rb]: Here you are calling `sanitize_column()` but it looks like it is no longer defined anywhere

File: Gemfile (hunk 3 above)
Comment (antonivanopoulos) [inline Gemfile]: Indeed haha, nice one :+1:

File: lib/paginative/models/model_extension.rb [hunk 5]
```diff
@@ -23,24 +26,88 @@ def self.with_id_from(id=0, limit=25)
...
-            # You can now pass in an array of 'field' params so that you can have a secondary sort order.
-            # This is important if your primary sort field could have duplicate values
-            primary_sort_field = field[0]
-            primary_value = value[0]
-            secondary_sort_field = field[1]
-            secondary_value = value[1]
-            # This allows us to pass in 2 different sort columns and still paginate correctly.
-            return self.order(sanitized_ordering(self.table_name, field, order)).where("#{self.table_name}.#{primary_sort_field} <= ? AND (#{self.table_name}.#{primary_sort_field} != ? OR #{self.table_name}.#{secondary_sort_field} < ?)", primary_value, primary_value, secondary_value) if order.try(:downcase) == "desc"
-
-            self.order(sanitized_ordering(self.table_name, field, order)).where("#{self.table_name}.#{primary_sort_field} >= ? AND (#{self.table_name}.#{primary_sort_field} != ? OR #{self.table_name}.#{secondary_sort_field} > ?)", primary_value, primary_value, secondary_value)
```
Comment (tspacek) [inline lib/paginative/models/model_extension.rb]: How are the sort fields sanitised?

File: lib/paginative/models/model_extension.rb [hunk 6]
```diff
@@ -23,24 +26,88 @@ def self.with_id_from(id=0, limit=25)
...
+
+        fields = Array.wrap(field).flatten
+        values = Array.wrap(value).flatten
+        zipped = fields.zip(values)
+        fields, values = prune_fields(zipped).transpose
```
Comment (tspacek) [inline lib/paginative/models/model_extension.rb]: Ah, gotcha.

**Comment by tspacek** [comment]
Only thing I'd say is put the description of this PR in there somewhere as a comment, or in the readme, to explain why it's the way it is.

**Comment by antonivanopoulos** [comment]
@tspacek Yeah nice idea :+1:

Can probably just chuck above the definition for `allow_paginative_on` + the README can include a snippet about it.

File: README.md [hunk 7]
```diff
@@ -101,9 +101,65 @@ models.with_field_from("id", 1)
 
 This will return all models that have an ID greater than 1, ordered by ID. Any column can be passed in, and the results will automatically be ordered by that column.
 
+
+### Allowed Paginative Fields
+
+Introduced in `0.3.0`, we've added a class method, `allow_paginative_on`, that allows you to whitelist the fields that you want to open up pagination on in your queries using `with_field_from`. By default, *no fields* are open for pagination using `with_field_from`, so you'll need to specify the ones you want to use before you can paginate anything with that method.
+
+Fields can be specified as symbols, which get mapped to the table of the current class:
+```
```
Comment (RustComet) [inline README.md]: Fun Fact: 
you can use ```ruby to syntax highlight code blocks as ruby.

**Comment by RustComet** [comment]
@antonivanopoulos README looks good! :+1:

File: README.md (hunk 7 above)
Comment (antonivanopoulos) [inline README.md]: :astonished: BRB, rejigging my code blocks.

**Pull request merged**